src/operators/InterestDistributor.sol

```diff
@@ -0,0 +1,103 @@
+///         distribute() should be called to update the outstanding shares that the interest is computed based on.
+///
+///         Uses peak price to calculate interest. If the price goes down, interest will only be redeemed
+///         again once the price fully recovers above the previous high point (peak).
+///         Peak is stored per user since if the price has globally gone down, and a user invests at that time,
+///         they'd expect to redeem interest based on the price they invested in, not the previous high point.
+/// @dev    Requires that:
+///         - Whenever orders are fulfilled, an UpdateTranchePrice message is also submitted
+///         - Users claim the complete fulfilled amount using the router, rather than partial claims through the vault.
+///         Otherwise the interest amounts may be off.
+contract InterestDistributor is IInterestDistributor {
+    using MathLib for uint256;
+
+    mapping(address vault => mapping(address user => InterestDetails)) internal _users;
+
+    IPoolManager public immutable poolManager;
+
+    constructor(address poolManager_) {
+        poolManager = IPoolManager(poolManager_);
+    }
+
+    /// @inheritdoc IInterestDistributor
+    function distribute(address vault, address controller) external {
+        IERC7540Vault vault_ = IERC7540Vault(vault);
+        require(vault_.isOperator(controller, address(this)), "InterestDistributor/not-an-operator");
+
+        InterestDetails storage user = _users[vault][controller];
+        uint128 prevShares = user.shares;
+
+        (address asset,) = poolManager.getVaultAsset(vault);
+        (uint128 currentPrice, uint64 priceLastUpdated) =
+            poolManager.getTranchePrice(vault_.poolId(), vault_.trancheId(), asset);
+        uint128 currentShares = IERC20(vault_.share()).balanceOf(controller).toUint128();
+
+        // Calculate request before updating user.shares, so it is based on the balance at the last price update.
+        // Assuming price updates coincide with epoch fulfillments, this results in only requesting
+        // interest on the previous outstanding balance before the new fulfillment.
+        uint128 request = priceLastUpdated > user.lastUpdate
+            ? _computeRequest(user.shares, currentShares, user.peak, uint96(currentPrice))
+            : 0;
+
+        user.lastUpdate = uint32(priceLastUpdated);
+        if (currentPrice > user.peak) user.peak = uint96(currentPrice);
+        user.shares = currentShares - request;
+
+        if (request > 0) {
+            vault_.requestRedeem(request, controller, controller);
+            emit InterestRedeemRequest(vault, controller, user.peak, uint96(currentPrice), request);
+        }
+
+        if (user.shares != prevShares) {
+            emit OutstandingSharesUpdate(vault, controller, prevShares, user.shares);
+        }
+    }
+
+    /// @inheritdoc IInterestDistributor
+    function clear(address vault, address controller) external {
```

Comment:
I would rename this to `dropOut`. Meaning that you want to drop-out from the interest distribution process.